trie/zktrie_deletionproof.go

```diff
@@ -0,0 +1,163 @@
+package trie
+
+import (
+	"bytes"
+
+	zktrie "github.com/scroll-tech/zktrie/trie"
+	zkt "github.com/scroll-tech/zktrie/types"
+
+	"github.com/scroll-tech/go-ethereum/ethdb"
+)
+
+// Pick Node from its hash directly from database, notice it has different
+// interface with the function of same name in `trie`
+func (t *ZkTrie) TryGetNode(nodeHash *zkt.Hash) (*zktrie.Node, error) {
+	if bytes.Equal(nodeHash[:], zkt.HashZero[:]) {
+		return zktrie.NewEmptyNode(), nil
+	}
+	nBytes, err := t.db.Get(nodeHash[:])
+	if err == zktrie.ErrKeyNotFound {
+		return nil, zktrie.ErrKeyNotFound
+	} else if err != nil {
+		return nil, err
+	}
+	return zktrie.NewNodeFromBytes(nBytes)
+}
+
+type ProofTracer struct {
+	*ZkTrie
+	deletionTracer map[zkt.Hash]struct{}
+	rawPaths       map[string][]*zktrie.Node
+}
+
+// NewProofTracer create a proof tracer object
+func (t *ZkTrie) NewProofTracer() *ProofTracer {
+	return &ProofTracer{
+		ZkTrie: t,
+		// always consider 0 is "deleted"
+		deletionTracer: map[zkt.Hash]struct{}{zkt.HashZero: {}},
+		rawPaths:       make(map[string][]*zktrie.Node),
+	}
+}
+
+// Merge merge the input tracer into current and return current tracer
+func (t *ProofTracer) Merge(another *ProofTracer) *ProofTracer {
+
+	// sanity checking
+	if !bytes.Equal(t.Hash().Bytes(), another.Hash().Bytes()) {
+		panic("can not merge two proof tracer base on different trie")
+	}
+
+	for k := range another.deletionTracer {
+		t.deletionTracer[k] = struct{}{}
```

Comment:
Why do we use the empty struct here?

Reply:
Just some common trick when we wish to use a set instead of map in golang: we put an empty struct as the value of a map so this map is worked as a set.